core/src/cluster_nodes.rs

```diff
@@ -0,0 +1,441 @@
+    pub fn get_broadcast_peer(&self, shred_seed: [u8; 32]) -> Option<&ContactInfo> {
+        if self.index.is_empty() {
+            None
+        } else {
+            let index = weighted_best(&self.index, shred_seed);
+            match &self.nodes[index].node {
+                NodeId::ContactInfo(node) => Some(node),
+                NodeId::Pubkey(_) => panic!("this should not happen!"),
+            }
+        }
+    }
+}
+
+impl ClusterNodes<RetransmitStage> {
+    pub fn new(cluster_info: &ClusterInfo, stakes: &HashMap<Pubkey, u64>) -> Self {
+        new_cluster_nodes(cluster_info, stakes)
+    }
+
+    pub fn get_retransmit_peers(
+        &self,
+        shred_seed: [u8; 32],
+        fanout: usize,
+        slot_leader: Option<Pubkey>,
+    ) -> (
+        Vec<&ContactInfo>, // neighbors
+        Vec<&ContactInfo>, // children
+    ) {
+        // Exclude leader from list of nodes.
+        let index = self.index.iter().copied();
+        let (weights, index): (Vec<u64>, Vec<usize>) = match slot_leader {
+            None => {
+                error!("unknown leader for shred slot");
+                index.unzip()
+            }
+            Some(slot_leader) if slot_leader == self.pubkey => {
+                error!("retransmit from slot leader: {}", slot_leader);
+                index.unzip()
+            }
+            Some(slot_leader) => index
+                .filter(|(_, i)| self.nodes[*i].pubkey() != slot_leader)
+                .unzip(),
+        };
+        let index: Vec<_> = {
+            let shuffle = weighted_shuffle(&weights, shred_seed);
+            shuffle.into_iter().map(|i| index[i]).collect()
+        };
+        let self_index = index
+            .iter()
+            .position(|i| self.nodes[*i].pubkey() == self.pubkey)
+            .unwrap();
+        let (neighbors, children) = compute_retransmit_peers(fanout, self_index, &index);
+        // Assert that the node itself is included in the set of neighbors, at
+        // the right offset.
+        debug_assert_eq!(
+            self.nodes[neighbors[self_index % fanout]].pubkey(),
+            self.pubkey
+        );
+        let get_contact_infos = |index: Vec<usize>| -> Vec<&ContactInfo> {
+            index
+                .into_iter()
+                .map(|i| self.nodes[i].contact_info().unwrap())
+                .collect()
+        };
+        (get_contact_infos(neighbors), get_contact_infos(children))
+    }
+}
+
+fn new_cluster_nodes<T: 'static>(
+    cluster_info: &ClusterInfo,
+    stakes: &HashMap<Pubkey, u64>,
+) -> ClusterNodes<T> {
+    let self_pubkey = cluster_info.id();
+    let nodes = get_nodes(cluster_info, stakes);
+    let broadcast = TypeId::of::<T>() == TypeId::of::<BroadcastStage>();
+    // For backward compatibility:
+    //   * nodes which do not have contact-info are excluded.
+    //   * stakes are floored at 1.
+    // The sorting key here should be equivalent to
+    // ClusterInfo::sorted_stakes_with_index.
```

Comment:
nit: equivalent to `solana_gossip::deprecated::sorted_stakes_with_index`

Reply:
done.